### PR TITLE
Temporary fix for AST visitor reentrancy bugs.

### PR DESCRIPTION
**Pull request by schweitzpgi**
This PR covers the first item in https://github.com/NVIDIA/cuda-quantum/issues/57.
The changes here will be validated by updating the CI pipelines as captured by subsequent items in that issue.

### CODE REVIEW

**Review by amccaskey — APPROVED** [review]

**Review by bettinaheim — CHANGES_REQUESTED** [review]
The requested changes are adding regression testss.

File: lib/Frontend/nvqpp/ConvertType.cpp [hunk 1]
```diff
@@ -440,13 +440,12 @@ SmallVector<Type> QuakeBridgeVisitor::lastTypes(unsigned n) {
 
 Type QuakeBridgeVisitor::genType(const clang::QualType &ty) {
   LLVM_DEBUG(llvm::dbgs() << "type to generate: " << ty << '\n');
-  bool saveTypeMode = typeMode;
-  typeMode = true;
-  bool res = TraverseType(ty);
+  QuakeBridgeVisitor vis(*this);
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply lib/Frontend/nvqpp/ConvertType.cpp:442]: Ok, let's track this under switching to libstdc++-12 then. 
Are the current tests covering the failure we saw, i.e. are the current tests failing without this fix?

I created an issue for the libstdc++ 12 update here: https://github.com/NVIDIA/cuda-quantum/issues/57. This PR should take care of the first item in the task list there. I can take care of the rest, if nothing else in the compiler needs to change. Let's switch to gcc12 then for the pipeline - this switch should be relevant only for us, and not matter for using CUDA Quantum, such that we don't need a deprecation strategy for 11.

**Comment by bettinaheim** [comment]
@schweitzpgi One more question: Could you clarify what exactly is "temporary" about the fix and should change in the future?

**Comment by schweitzpgi** [comment]
> @schweitzpgi One more question: Could you clarify what exactly is "temporary" about the fix and should change in the future?

I should write up a RFC. Longer term, we want the bridge to be able to construct and use types as part of the traversal (single pass). At the moment, there are two semi-independent and not necessarily related traversals happening and the type being generated (hopefully) agrees with the code generated. So the current state is awkward.

**Pull request merged**